src/coreclr/interpreter/compiler.cpp

```diff
@@ -33,7 +34,56 @@ static const InterpType g_interpTypeFromStackType[] =
 
 static const char *g_stackTypeString[] = { "I4", "I8", "R4", "R8", "O ", "VT", "MP", "F " };
 
-// FIXME Use specific allocators for their intended purpose
+/*****************************************************************************/
+void DECLSPEC_NORETURN Interp_NOMEM()
+{
+    assert(false);
+    // Ensure we don't return and that the compiler knows we won't
+#ifdef _MSC_VER
```

Comment:
This is what we use in coreclr runtime:
```c++
#if defined(_MSC_VER) || defined(_PREFIX_)
#define __UNREACHABLE() __assume(0)
#else
#define __UNREACHABLE() __builtin_unreachable()
#endif
```

Reply:
That's helpful, thanks! Just to be clear, does `__assume(0)` halt execution? I think in this case we don't want `__builtin_unreachable` because this function is for an out-of-memory condition, which can happen and should result in some sort of orderly abort or runtime exception.

I will need to do something similar to replace the use of _DbgBreak in the assert macros since that apparently doesn't work on clang.

Reply:
`__assume(0)` tells the compiler that this is guaranteed to be unreachable code path. Optimizations are free to turn `__assume(0)` into anything - no-op, breakpoint, crash, infinite loop, ... .

Reply:
Should I just throw a C++ exception? Or do we not do that in this part of the codebase?

Reply:
Thanks. I will look at what RyuJIT does for similar scenarios for reference.

Reply:
The JIT appears to use RaiseException from the PAL for this purpose; my understanding is that we can't use the PAL in the interpreter. I suppose I can just throw a bare exception though, as that appears to be the PAL implementation of RaiseException in some cases.

Reply:
Yes, you can just use standard C++ OOM exception.